File: supabase/crisisAnalysis.py
import pickle
import numpy as np
import pandas as pd
import shap
from typing import Dict
import os
import warnings
import logging

warnings.filterwarnings('ignore')

# Setup logging
logging.basicConfig(
    level=logging.DEBUG,
    format='%(asctime)s - %(levelname)s - %(message)s'
)
logger = logging.getLogger(__name__)

# =====================================================================
# Load NEW Pickle-Based Models
# =====================================================================
economic_model_data = None
food_model_data = None

# Try to load ECONOMIC model
economic_paths = [
    'models/economic_model.pkl',
    './models/economic_model.pkl',
    'backend/models/economic_model.pkl',
]

for path in economic_paths:
    if os.path.exists(path):
        try:
            with open(path, 'rb') as f:
                economic_model_data = pickle.load(f)
            
            logger.info(f"✅ Economic model loaded from: {path}")
            
            # Verify structure
            if isinstance(economic_model_data, dict):
                logger.info(f"   Model type: {type(economic_model_data.get('model'))}")
                logger.info(f"   Feature count: {len(economic_model_data.get('feature_order', []))}")
                logger.info(f"   Features: {economic_model_data.get('feature_order')}")
            else:
                logger.warning(f"   ⚠️ Model is not a dict, might be old format")
            
            break
        except Exception as e:
            logger.warning(f"⚠️ Failed to load from {path}: {e}")

# Try to load FOOD model
food_paths = [
    'models/food_model.pkl',
    './models/food_model.pkl',
    'backend/models/food_model.pkl',
]

for path in food_paths:
    if os.path.exists(path):
        try:
            with open(path, 'rb') as f:
                food_model_data = pickle.load(f)
            
            logger.info(f"✅ Food model loaded from: {path}")
            
            # Verify structure
            if isinstance(food_model_data, dict):
                logger.info(f"   Model type: {type(food_model_data.get('model'))}")
                logger.info(f"   Feature count: {len(food_model_data.get('feature_order', []))}")
                logger.info(f"   Features: {food_model_data.get('feature_order')}")
            else:
                logger.warning(f"   ⚠️ Model is not a dict, might be old format")
            
            break
        except Exception as e:
            logger.warning(f"⚠️ Failed to load from {path}: {e}")

# =====================================================================
# ECONOMIC CRISIS ANALYSIS - NEW VERSION (NO LAGS)
# =====================================================================

def analyze_economic_crisis(
    country: str,
    gdp_growth: float,
    inflation: float,
    unemployment: float,
    domestic_credit: float,
    exports: float,
    imports: float,
    gdp_per_capita: float = 0.0,  # NEW - required by new model
    gross_fixed_capital: float = 0.0,  # NEW - required by new model
) -> Dict:
    """
    Economic crisis analysis using NEW pickle-based model
    
    NEW MODEL FEATURES (8 total, NO lags):
    1. Domestic credit to private sector (% of GDP)
    2. Exports of goods and services (% of GDP)
    3. GDP growth (annual %)
    4. GDP per capita (current US$)
    5. Gross fixed capital formation (% of GDP)
    6. Imports of goods and services (% of GDP)
    7. Inflation, consumer prices (annual %)
    8. Unemployment, total (% of total labor force)
    """
    
    if economic_model_data is None:
        logger.error("Economic model not loaded!")
        return {
            'country': country,
            'probability': 0,
            'classification': 'Error: Model not loaded',
            'topIndicators': [],
        }

    try:
        # Extract model and feature order
        model = economic_model_data.get('model')
        feature_order = economic_model_data.get('feature_order', [])
        
        if model is None or not feature_order:
            logger.error("Model or feature_order missing from pickle!")
            return {
                'country': country,
                'probability': 0,
                'classification': 'Error: Invalid model structure',
                'topIndicators': [],
            }

        logger.debug(f"Economic Analysis for {country}:")
        logger.debug(f"  GDP Growth: {gdp_growth}")
        logger.debug(f"  Inflation: {inflation}")
        logger.debug(f"  Unemployment: {unemployment}")
        logger.debug(f"  Domestic Credit: {domestic_credit}")
        logger.debug(f"  Exports: {exports}")
        logger.debug(f"  Imports: {imports}")
        logger.debug(f"  GDP per Capita: {gdp_per_capita}")
        logger.debug(f"  Gross Fixed Capital: {gross_fixed_capital}")

        # Create feature dictionary matching training data
        features_dict = {
            'Domestic credit to private sector (% of GDP)': domestic_credit,
            'Exports of goods and services (% of GDP)': exports,
            'GDP growth (annual %)': gdp_growth,
            'GDP per capita (current US$)': gdp_per_capita,
            'Gross fixed capital formation (% of GDP)': gross_fixed_capital,
            'Imports of goods and services (% of GDP)': imports,
            'Inflation, consumer prices (annual %)': inflation,
            'Unemployment, total (% of total labor force) (modeled ILO estimate)': unemployment,
        }

        # Build feature vector in EXACT order from training
        feature_vector = np.array([features_dict[feat] for feat in feature_order]).reshape(1, -1)

        logger.debug(f"Feature vector shape: {feature_vector.shape}")
        logger.debug(f"Feature vector: {feature_vector}")

        # Verify feature count
        if hasattr(model, 'n_features_in_'):
            expected = model.n_features_in_
            actual = feature_vector.shape[1]
            if expected != actual:
                logger.error(f"❌ FEATURE MISMATCH! Expected {expected}, got {actual}")
                return {
                    'country': country,
                    'probability': 0,
                    'classification': 'Error: Feature count mismatch',
                    'topIndicators': [],
                }

        # Predict probability
        if hasattr(model, 'predict_proba'):
            probability_raw = model.predict_proba(feature_vector)[0][1]
        else:
            probability_raw = model.predict(feature_vector)[0]

        probability = probability_raw * 100
        logger.info(f"Final probability: {probability:.2f}% (raw: {probability_raw:.4f})")

        # Classification
        if probability >= 80:
            classification = 'High Risk'
        elif probability >= 50:
            classification = 'Moderate Risk'
        else:
            classification = 'Low Risk'

        # SHAP values for top indicators
        try:
            explainer = shap.TreeExplainer(model)
            shap_values = explainer.shap_values(feature_vector)

            if isinstance(shap_values, list) and len(shap_values) > 1:
                shap_positive = shap_values[1]
            else:
                shap_positive = shap_values

            if hasattr(shap_positive, 'ndim') and shap_positive.ndim == 3:
                shap_positive = shap_positive[:, :, 1]

            instance_shap = shap_positive[0]

            # Pair with actual values
            feature_contribs = sorted(
                zip(feature_order, instance_shap, feature_vector[0]),
                key=lambda x: abs(x[1]),
                reverse=True
            )[:3]

            top_indicators = [
                {
                    'name': name,
                    'impact': float(contrib),
                    'value': float(val)
                }
                for name, contrib, val in feature_contribs
            ]

        except Exception as shap_error:
            logger.warning(f"⚠️ SHAP failed: {shap_error}")
            top_indicators = []

        result = {
            'country': country,
            'probability': round(probability, 2),
            'classification': classification,
            'topIndicators': top_indicators,
            'isHighRisk': bool(probability >= 80)
        }

        logger.info(f"✅ {country}: {probability:.2f}% - {classification}")
        return result

    except Exception as e:
        logger.error(f"❌ Error analyzing {country}: {str(e)}")
        import traceback
        logger.error(traceback.format_exc())
        return {
            'country': country,
            'probability': 0,
            'classification': 'Error',
            'topIndicators': [],
        }

# =====================================================================
# FOOD CRISIS ANALYSIS - NEW VERSION (NO LAGS)
# =====================================================================

def analyze_food_crisis(
    country: str,
    cereal_yield: float,
    food_imports: float,
    food_production_index: float,
    gdp_growth: float,
    gdp_per_capita: float,
    inflation: float,
    population_growth: float,
    gdp_current: float = 0.0,  # NEW - required by new model
) -> Dict:
    """
    Food crisis analysis using NEW pickle-based model
    
    NEW MODEL FEATURES (8 total, NO lags):
    1. Cereal yield (kg per hectare)
    2. Food imports (% of merchandise imports)
    3. Food production index (2014-2016 = 100)
    4. GDP (current US$)  ← NEW!
    5. GDP growth (annual %)
    6. GDP per capita (current US$)
    7. Inflation, consumer prices (annual %)
    8. Population growth (annual %)
    """
    
    if food_model_data is None:
        logger.error("Food model not loaded!")
        return {
            'country': country,
            'probability': 0,
            'classification': 'Error: Model not loaded',
            'topIndicators': [],
        }

    try:
        # Extract model and feature order
        model = food_model_data.get('model')
        feature_order = food_model_data.get('feature_order', [])
        
        if model is None or not feature_order:
            logger.error("Model or feature_order missing from pickle!")
            return {
                'country': country,
                'probability': 0,
                'classification': 'Error: Invalid model structure',
                'topIndicators': [],
            }

        logger.debug(f"Food Analysis for {country}:")
        logger.debug(f"  Cereal Yield: {cereal_yield}")
        logger.debug(f"  Food Imports: {food_imports}")
        logger.debug(f"  Food Production Index: {food_production_index}")
        logger.debug(f"  GDP Growth: {gdp_growth}")
        logger.debug(f"  GDP Per Capita: {gdp_per_capita}")
        logger.debug(f"  Inflation: {inflation}")
        logger.debug(f"  Population Growth: {population_growth}")
        logger.debug(f"  GDP Current: {gdp_current}")

        # Create feature dictionary
        features_dict = {
            'Cereal yield (kg per hectare)': cereal_yield,
            'Food imports (% of merchandise imports)': food_imports,
            'Food production index (2014-2016 = 100)': food_production_index,
            'GDP (current US$)': gdp_current,
            'GDP growth (annual %)': gdp_growth,
            'GDP per capita (current US$)': gdp_per_capita,
            'Inflation, consumer prices (annual %)': inflation,
            'Population growth (annual %)': population_growth,
        }

        # Build feature vector in EXACT order from training
        feature_vector = np.array([features_dict[feat] for feat in feature_order]).reshape(1, -1)

        logger.debug(f"Feature vector shape: {feature_vector.shape}")
        logger.debug(f"Feature vector: {feature_vector}")

        # Verify feature count
        if hasattr(model, 'n_features_in_'):
            expected = model.n_features_in_
            actual = feature_vector.shape[1]
            if expected != actual:
                logger.error(f"❌ FEATURE MISMATCH! Expected {expected}, got {actual}")
                return {
                    'country': country,
                    'probability': 0,
                    'classification': 'Error: Feature count mismatch',
                    'topIndicators': [],
                }

        # Predict probability
        if hasattr(model, 'predict_proba'):
            probability_raw = model.predict_proba(feature_vector)[0][1]
        else:
            probability_raw = model.predict(feature_vector)[0]

        probability = probability_raw * 100
        logger.info(f"Final probability: {probability:.2f}% (raw: {probability_raw:.4f})")

        # Classification
        if probability >= 85:
            classification = 'Critical'
        elif probability >= 70:
            classification = 'High Risk'
        elif probability >= 50:
            classification = 'Moderate Risk'
        else:
            classification = 'Low Risk'

        # SHAP values
        try:
            explainer = shap.TreeExplainer(model)
            shap_values = explainer.shap_values(feature_vector)

            if isinstance(shap_values, list) and len(shap_values) > 1:
                shap_positive = shap_values[1]
            else:
                shap_positive = shap_values

            if hasattr(shap_positive, 'ndim') and shap_positive.ndim == 3:
                shap_positive = shap_positive[:, :, 1]

            instance_shap = shap_positive[0]

            feature_contribs = sorted(
                zip(feature_order, instance_shap, feature_vector[0]),
                key=lambda x: abs(x[1]),
                reverse=True
            )[:3]

            top_indicators = [
                {
                    'name': name,
                    'impact': float(contrib),
                    'value': float(val)
                }
                for name, contrib, val in feature_contribs
            ]

        except Exception as shap_error:
            logger.warning(f"⚠️ SHAP failed: {shap_error}")
            top_indicators = []

        result = {
            'country': country,
            'probability': round(probability, 2),
            'classification': classification,
            'topIndicators': top_indicators,
            'isHighRisk': bool(probability >= 70)
        }

        logger.info(f"✅ {country}: {probability:.2f}% - {classification}")
        return result

    except Exception as e:
        logger.error(f"❌ Error analyzing {country}: {str(e)}")
        import traceback
        logger.error(traceback.format_exc())
        return {
            'country': country,
            'probability': 0,
            'classification': 'Error',
            'topIndicators': [],
        }

File: src/lib/crisisAnalysis.ts
export interface EconomicInput {
  gdpGrowth: number;
  inflation: number;
  unemployment: number;
  domesticCredit: number;
  exports: number;
  imports: number;
  gdpPerCapita: number;           // NEW - Required
  grossFixedCapital: number;      // NEW - Required
}

export interface FoodInput {
  cerealYield: number;
  foodImports: number;
  foodProductionIndex: number;
  gdpGrowth: number;
  gdpPerCapita: number;
  inflation: number;
  populationGrowth: number;
  gdpCurrent: number;             // NEW - Required (total GDP in US$, NOT per capita)
}

export interface Indicator {
  name: string;
  impact: number;
  value: number;
}

export interface CrisisResult {
  country: string;
  probability: number;
  classification: string;
  topIndicators: Indicator[];
  isHighRisk?: boolean;
}

// =====================================================================
// Backend URL Configuration (use environment variables)
// =====================================================================

const API_BASE_URL = import.meta.env.VITE_API_URL || 'http://localhost:3001';

// =====================================================================
// VALIDATION HELPER
// =====================================================================

interface ValidationResult {
  isValid: boolean;
  errors: string[];
}

function validateEconomicInput(data: EconomicInput): ValidationResult {
  const errors: string[] = [];

  // Required fields
  if (!Number.isFinite(data.gdpGrowth)) errors.push('gdpGrowth must be a valid number');
  if (!Number.isFinite(data.inflation)) errors.push('inflation must be a valid number');
  if (!Number.isFinite(data.unemployment)) errors.push('unemployment must be a valid number');
  if (!Number.isFinite(data.domesticCredit)) errors.push('domesticCredit must be a valid number');
  if (!Number.isFinite(data.exports)) errors.push('exports must be a valid number');
  if (!Number.isFinite(data.imports)) errors.push('imports must be a valid number');
  if (!Number.isFinite(data.gdpPerCapita)) errors.push('gdpPerCapita must be a valid number');
  if (!Number.isFinite(data.grossFixedCapital)) errors.push('grossFixedCapital must be a valid number');

  return {
    isValid: errors.length === 0,
    errors
  };
}

function validateFoodInput(data: FoodInput): ValidationResult {
  const errors: string[] = [];

  // Required fields
  if (!Number.isFinite(data.cerealYield)) errors.push('cerealYield must be a valid number');
  if (!Number.isFinite(data.foodImports)) errors.push('foodImports must be a valid number');
  if (!Number.isFinite(data.foodProductionIndex)) errors.push('foodProductionIndex must be a valid number');
  if (!Number.isFinite(data.gdpGrowth)) errors.push('gdpGrowth must be a valid number');
  if (!Number.isFinite(data.gdpPerCapita)) errors.push('gdpPerCapita must be a valid number');
  if (!Number.isFinite(data.inflation)) errors.push('inflation must be a valid number');
  if (!Number.isFinite(data.populationGrowth)) errors.push('populationGrowth must be a valid number');
  if (!Number.isFinite(data.gdpCurrent)) errors.push('gdpCurrent must be a valid number');

  return {
    isValid: errors.length === 0,
    errors
  };
}

// =====================================================================
// Economic Crisis Analysis - PICKLE VERSION (NO LAGS)
// =====================================================================

export async function analyzeEconomicCrisis(
  country: string,
  data: EconomicInput
): Promise<CrisisResult> {
  try {
    // Validate input data
    const validation = validateEconomicInput(data);
    if (!validation.isValid) {
      console.error('Invalid economic input:', validation.errors);
      return {
        country,
        probability: 0,
        classification: 'Error: Invalid input data',
        topIndicators: [],
      };
    }

    // Prepare request payload (NO LAG FEATURES)
    const payload = {
      country,
      gdpGrowth: data.gdpGrowth,
      inflation: data.inflation,
      unemployment: data.unemployment,
      domesticCredit: data.domesticCredit,
      exports: data.exports,
      imports: data.imports,
      gdpPerCapita: data.gdpPerCapita,           // NEW
      grossFixedCapital: data.grossFixedCapital, // NEW
    };

    console.log('📊 Sending Economic Crisis Analysis Request:', {
      country,
      gdpGrowth: data.gdpGrowth,
      inflation: data.inflation,
      gdpPerCapita: data.gdpPerCapita,
      grossFixedCapital: data.grossFixedCapital,
    });

    const response = await fetch(`${API_BASE_URL}/api/analyze/economic`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(payload),
    });

    if (!response.ok) {
      const errorText = await response.text();
      console.error(`API error (${response.status}):`, errorText);
      throw new Error(`API error: ${response.status} ${response.statusText}`);
    }

    const result = await response.json();
    console.log('✅ Economic Analysis Result:', result);

    return result;

  } catch (error) {
    console.error('❌ Economic crisis analysis error:', error);
    return {
      country,
      probability: 0,
      classification: 'Error: Analysis failed',
      topIndicators: [],
    };
  }
}

// =====================================================================
// Food Crisis Analysis - PICKLE VERSION (NO LAGS)
// =====================================================================

export async function analyzeFoodCrisis(
  country: string,
  data: FoodInput
): Promise<CrisisResult> {
  try {
    // Validate input data
    const validation = validateFoodInput(data);
    if (!validation.isValid) {
      console.error('Invalid food input:', validation.errors);
      return {
        country,
        probability: 0,
        classification: 'Error: Invalid input data',
        topIndicators: [],
      };
    }

    // Prepare request payload (NO LAG FEATURES)
    const payload = {
      country,
      cerealYield: data.cerealYield,
      foodImports: data.foodImports,
      foodProductionIndex: data.foodProductionIndex,
      gdpGrowth: data.gdpGrowth,
      gdpPerCapita: data.gdpPerCapita,
      inflation: data.inflation,
      populationGrowth: data.populationGrowth,
      gdpCurrent: data.gdpCurrent,  // NEW
    };

    console.log('🍽️ Sending Food Crisis Analysis Request:', {
      country,
      cerealYield: data.cerealYield,
      foodImports: data.foodImports,
      foodProductionIndex: data.foodProductionIndex,
      gdpCurrent: data.gdpCurrent,
    });

    const response = await fetch(`${API_BASE_URL}/api/analyze/food`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(payload),
    });

    if (!response.ok) {
      const errorText = await response.text();
      console.error(`API error (${response.status}):`, errorText);
      throw new Error(`API error: ${response.status} ${response.statusText}`);
    }

    const result = await response.json();
    console.log('✅ Food Analysis Result:', result);

    return result;

  } catch (error) {
    console.error('❌ Food crisis analysis error:', error);
    return {
      country,
      probability: 0,
      classification: 'Error: Analysis failed',
      topIndicators: [],
    };
  }
}

// =====================================================================
// Health Check
// =====================================================================

export async function checkAPIHealth(): Promise<{
  isHealthy: boolean;
  details: {
    apiUrl: string;
    status: string;
    timestamp: string;
  };
}> {
  try {
    const response = await fetch(`${API_BASE_URL}/health`);

    return {
      isHealthy: response.ok,
      details: {
        apiUrl: API_BASE_URL,
        status: response.ok ? 'OK' : 'ERROR',
        timestamp: new Date().toISOString()
      }
    };
  } catch (error) {
    console.error('❌ Health check failed:', error);
    return {
      isHealthy: false,
      details: {
        apiUrl: API_BASE_URL,
        status: 'UNREACHABLE',
        timestamp: new Date().toISOString()
      }
    };
  }
}